fungible_asset_wrapper/sources/fungible_asset_wrapper.move

```diff
@@ -0,0 +1,168 @@
+/// Anyone can call this module to permissionlessly wrap a v1 coin into a v2 coin, and unwrap vice versa.
+module fungible_asset_wrapper::fungible_asset_wrapper {
+    use std::option;
+    use std::signer;
+    use std::string;
+
+    use aptos_std::type_info;
+    use aptos_framework::account::{Self, SignerCapability};
+    use aptos_framework::coin::{Self, Coin};
+    use aptos_framework::fungible_asset::{Self, FungibleAsset, MintRef, BurnRef};
+    use aptos_framework::object;
+    use aptos_framework::primary_fungible_store;
+    use aptos_framework::resource_account;
+
+    #[test_only]
+    use aptos_framework::coin::{MintCapability, BurnCapability};
+    #[test_only]
+    use aptos_framework::fungible_asset::Metadata;
+
+    const RESOURCE_ACCOUNT_ADDRESS: address = @fungible_asset_wrapper;
+    const DEPLOYER_ADDRESS: address = @deployer;
+    const RESOURCE_ACCOUNT_SEED: vector<u8> = b"fungible_asset_wrapper";
+
+    struct SignerCapBox has key {
+        signer_cap: SignerCapability
+    }
+
+    struct FungibleAssetAdmin has key {
+        mint_ref: MintRef,
+        burn_ref: BurnRef
+    }
+
+    fun init_module(resource_signer: &signer) {
+        let signer_cap = resource_account::retrieve_resource_account_cap(resource_signer, DEPLOYER_ADDRESS);
+        move_to(resource_signer, SignerCapBox { signer_cap });
+    }
+
+    public entry fun register<CoinType>() acquires SignerCapBox {
+        let resource_account_signer = resource_account_signer();
+        let constructor_ref = object::create_named_object(
+            &resource_account_signer,
+            *string::bytes(&type_info::type_name<CoinType>())
+        );
+        primary_fungible_store::create_primary_store_enabled_fungible_asset(
+            &constructor_ref,
+            option::none(),
+            coin::name<CoinType>(),
+            coin::symbol<CoinType>(),
+            coin::decimals<CoinType>(),
+            string::utf8(b""),
+            string::utf8(b""),
+        );
+        move_to(
+            &object::generate_signer(&constructor_ref),
+            FungibleAssetAdmin {
+                mint_ref: fungible_asset::generate_mint_ref(&constructor_ref),
+                burn_ref: fungible_asset::generate_burn_ref(&constructor_ref)
+            }
+        );
+        if (!coin::is_account_registered<CoinType>(signer::address_of(&resource_account_signer))) {
+            coin::register<CoinType>(&resource_account_signer);
+        }
+    }
+
+    public fun wrap<CoinType>(coin: Coin<CoinType>): FungibleAsset acquires FungibleAssetAdmin {
+        let amount = coin::value(&coin);
+        let coin_address = coin_address<CoinType>();
+        let fa_admin = borrow_global<FungibleAssetAdmin>(coin_address);
+        coin::deposit(resource_account_address(), coin);
```

Comment:
This seems kinda like a rugpull. You wrap your coins using my module and I'll steal all your money

Reply:
@movekevin but you get a fungible asset version of your coin back 😆 thought it's fair

Reply:
Yeah but that would not be backed by anything if all underlying tokens are withdrawn :)

Reply:
i think all minted fungible assets are backed by v1 coins held by the resource account.
also, i make this module intentionally immutable so that people cannot add a `withdraw_coin` method to pull out the backing v1 coins.
@movekevin does that sound good? if not what's your suggestion here?